netmon/netmon.go

```diff
@@ -0,0 +1,613 @@
+	kataCLIUpdtRoutesCmd = "update-routes"
+
+	kataSuffix = "kata"
+
+	// For simplicity the code will only focus on IPv4 addresses for now.
+	netlinkFamily = netlink.FAMILY_V4
+
+	storageParentPath = "/var/run/kata-containers/netmon/sbs"
+	storageDirPerm    = 0750
+
+	// sharedFile is the name of the file that will be used to share
+	// the data between this process and the kata-runtime process
+	// responsible for updating the network.
+	sharedFile      = "shared.json"
+	storageFilePerm = 0640
+)
+
+// version is the netmon version. This variable is populated at build time.
+var version = "unknown"
+
+type netmonParams struct {
+	sandboxID   string
+	runtimePath string
+	debug       bool
+	logLevel    string
+}
+
+type netmon struct {
+	netmonParams
+
+	storagePath string
+	sharedFile  string
+
+	netIfaces map[int]Interface
+
+	linkUpdateCh chan netlink.LinkUpdate
+	linkDoneCh   chan struct{}
+
+	rtUpdateCh chan netlink.RouteUpdate
+	rtDoneCh   chan struct{}
+
+	netHandler *netlink.Handle
+}
+
+var netmonLog = logrus.New()
+
+func printVersion() {
+	fmt.Printf("%s version %s\n", netmonName, version)
+}
+
+func parseOptions() netmonParams {
+	var version bool
+
+	params := netmonParams{}
+
+	flag.BoolVar(&params.debug, "d", false, "enable debug mode")
+	flag.BoolVar(&version, "v", false, "display program version and exit")
+	flag.StringVar(&params.sandboxID, "s", "", "sandbox id (required)")
+	flag.StringVar(&params.runtimePath, "r", "", "runtime path (required)")
+	flag.StringVar(&params.logLevel, "log", "warn",
+		"log messages above specified level: debug, warn, error, fatal or panic")
+
+	flag.Parse()
+
+	if version {
+		printVersion()
+		os.Exit(0)
+	}
+
+	if params.sandboxID == "" {
+		fmt.Fprintf(os.Stderr, "Error: sandbox id is empty, one must be provided\n")
+		flag.PrintDefaults()
+		os.Exit(1)
+	}
+
+	if params.runtimePath == "" {
+		fmt.Fprintf(os.Stderr, "Error: runtime path is empty, one must be provided\n")
+		flag.PrintDefaults()
+		os.Exit(1)
+	}
+
+	return params
+}
+
+func newNetmon(params netmonParams) (*netmon, error) {
+	handler, err := netlink.NewHandle(netlinkFamily)
+	if err != nil {
+		return nil, err
+	}
+
+	n := &netmon{
+		netmonParams: params,
+		storagePath:  filepath.Join(storageParentPath, params.sandboxID),
+		sharedFile:   filepath.Join(storageParentPath, params.sandboxID, sharedFile),
+		netIfaces:    make(map[int]Interface),
+		linkUpdateCh: make(chan netlink.LinkUpdate),
+		linkDoneCh:   make(chan struct{}),
+		rtUpdateCh:   make(chan netlink.RouteUpdate),
+		rtDoneCh:     make(chan struct{}),
+		netHandler:   handler,
+	}
+
+	if err := os.MkdirAll(n.storagePath, storageDirPerm); err != nil {
+		return nil, err
+	}
+
+	return n, nil
+}
+
+func (n *netmon) cleanup() {
+	os.RemoveAll(n.storagePath)
+	n.netHandler.Delete()
+	close(n.linkDoneCh)
+	close(n.rtDoneCh)
+}
+
+func (n *netmon) logger() *logrus.Entry {
+	fields := logrus.Fields{
+		"name":   netmonName,
+		"pid":    os.Getpid(),
+		"source": "netmon",
```

Comment:
Nice. This should allow it to work well with https://github.com/kata-containers/tests/tree/master/cmd/log-parser.

But thinking about the common code in the other components, what about signal handling? It would be great if this too allowed `SIGUSR1` to dump a backtrace (https://github.com/kata-containers/documentation/blob/master/Developer-Guide.md#troubleshoot-kata-containers).

You can rework `cli/signals.go` like I did for the throttler to avoid code duplication:

- https://github.com/kata-containers/ksm-throttler/pull/41

Reply:
@jodh-intel in order to factorize the code, we need to define a new package, otherwise I don't see how I can share between the `netmon` and the `cli`. Also, I realized that the PR is pretty big now, and defining a new package and adding this backtrace support can definitely be done in a follow up PR. WDYT?

Reply:
Agree. Because cli is the main package, we are not able to reuse code in it.